presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -231,4 +233,9 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
                 .map(taskContext -> taskContext.accept(visitor, context))
                 .collect(toList());
     }
+
+    public Optional<TaskContext> getTaskContextByTaskId(TaskId taskId)
+    {
+        return taskContexts.stream().filter(context -> context.getTaskId().equals(taskId)).findFirst();
```

Comment:
That was my thinking but unfortunately I don't know what is the life cycle of tasks contexts within `QueryContext`. I saw the `CopyOnWriteArrayList` which alerts something is going on... However, from within `SqlTask` it looks like it should be safe to keep its own context.